display images, HTML... and MIME rendering

cosmos72 opened the pull request:
show as image each expression result that implements Go interface image.Image

dwhitena commented:
This is great @cosmos72. Thank you so much! I have been swamped, but I'm hoping to test this soon. @SpencerPark, have you been able to test this?

SpencerPark commented on display.go:
```diff
@@ -0,0 +1,116 @@
+// See http://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.display
+// for a good overview of the support types. Note: This is missing _repr_markdown_ and _repr_javascript_.
+
+var globalReceipt *msgReceipt // ugly global variable. any alternative?
```
I think restricting this to just rendering (returning a mime bundle) and pulling the actual display portion will clean this up a lot.

SpencerPark commented on display.go:
```diff
@@ -0,0 +1,116 @@
+			mimeType:     data,
+		}, make(bundledMIMEData))
+}
+
```
`render2` and `render3` would then return `map[string]interface{}` instead of publishing.

SpencerPark commented on image.go:
```diff
@@ -27,7 +27,7 @@ func publishImages(vals []interface{}, receipt msgReceipt) []interface{} {
+func publishImage(img image.Image, receipt *msgReceipt) error {
 	bytes, mime, err := encodePng(img)
 	if err != nil {
 		return err
```
These functions can move into the same package for rendering and ideally keep them more or less pure. This separation will be really helpful both for testing and maintaining.

SpencerPark commented on kernel.go:
```diff
@@ -95,6 +95,12 @@ func runKernel(connectionFile string) {
+	if err != nil {
+		log.Print(err)
+	}
+
```
Something like in f696d7632076934f53d928ec5fe4cee5c222fb64 that replaces the display function with a closure that knows about the eval context removes the global. Not sure of any repercussions gomacro will have as a result but taking advantage of the more dynamic interpreter seemed like a great use for gomacro.

SpencerPark commented on kernel.go:
```diff
@@ -359,7 +372,7 @@ func handleExecuteRequest(ir *interp.Interp, receipt msgReceipt) error {
 	if executionErr == nil {
 		// if one or more value is image.Image, display it instead
-		vals = publishImages(vals, receipt)
+		vals = publishImages(vals, &receipt)
```
If all of the functions simply render instead of display this is the place we get to take advantage of that. We need a type assertion but I think checking for a struct is fine. If we are trying to return a mimebundle we can get the rich output in the `Out[_]` in addition to the display calls.

SpencerPark commented:
I might have reviewed at the wrong commit... anyways I like the changes but I think the separation of rendering and displaying is really important. I also had some unit tests in the works including the helpers for it which I would like to bring over here. 

This is such a popular feature that we should really make sure to get something stable right off the bat.

cosmos72 commented:
You reviewed the right commit, don't worry.

Yes, I saw you injected into the interpreter closures that kept a reference to the current msgReceipt. At least syntactically, it's much cleaner than my global variable.

But the real question is: what's the lifetime of the msgReceipt passed to handleExecuteRequest() ?
Can it still be used after handleExecuteRequest() returns or not?

Closures on the msgReceipt will remain available to the interpreter even after Interp.RunExpr() returns - for example, one may attempt to display from a long-running goroutine. I have no idea what would happen in such case.
On the other hand, my patch sets globalReceipt to nil when handleExecuteRequest() returns. This prevents long-running interpreted goroutines from sending display requests to gophernotes and Jupyter. Clearly, a similar technique can be used also for closures.

About splitting the rendering (returning a mime bundle) from the actual display portion: I am not fond of exposing very complex API to interpreted code - that's also why I tried to simplify your implementation. The most complex function I exposed to the interpreter is 'display.MIME(data, metadata map[string]interface{}) error` and, while needed for the fully general case, it already feels complicated to use for me.
In any case, *some* code will have to send the Jupiter message containing the mime bundle, and such code will have the same problem of accessing the correct msgReceipt

SpencerPark commented:
>   But the real question is: what's the lifetime of the msgReceipt passed to handleExecuteRequest() ?
Can it still be used after handleExecuteRequest() returns or not?

Good point. I believe the behavior is undefined in the spec and so we can't rely on the server to do anything with it but I don't think there is any harm in using it. In either case it might be worth setting it to `nil` as you have done but maybe still keep it local (and hidden).

>   I am not fond of exposing very complex API to interpreted code - that's also why I tried to simplify your implementation. The most complex function I exposed to the interpreter is display.MIME(data, metadata map[string]interface{}) error and, while needed for the fully general case, it already feels complicated to use for me.

I do agree with this in most cases but the reason I am so strong on it is because of the very long running precedent that Jupyter has set with IPython's display mechanics. The _display_ function and it's behavior should essentially give access to some print stream for richer output in the same way that returning something (an expression at the end of a cell) should. This is why the rendering/display separation is required as something can be rendered and used as the ExecuteResult instead of DisplayData.

I'm fine with it being simplified to always return something like `DisplayData` and keep interfaces out of it but the user code needs access to these mechanics even if the rendering always needs to be explicit.

>   In any case, some code will have to send the Jupiter message containing the mime bundle, and such code will have the same problem of accessing the correct msgReceipt.

Hopefully we can hide as much of that as possible which I propose we do with a `Display` function taking some `DisplayData`.

cosmos72 commented:
Thanks, it's clearer now. I guess I should read IPython API...
If I understand correctly, you want an API like
```
type DisplayData struct { Data, Metadata, Transient map[string]interface{} } // or possibly something else to be defined
func Display(data DisplayData) error
```
plus some other utility functions that create such DisplayDaty objects,
plus extending gophernotes to detect any DisplayData object returned by the interpreter and send them to Jupiter for displaying?

That makes sense.

cosmos72 commented:
I just updated my branch (and as a consequence this pull request) to better follow your suggestion and IPython API, including the separation between the `func Display(data DisplayData) error` that sends multimedia data to Jupyter for displaying, and the the various factory functions 'func display.*(...) DisplayData' that create and return DisplayData values.

Since I was at it, I replaced the global variable with a closure, and also updated 'publishImages' to also detect (and show) DIsplayData values returned at prompt.

SpencerPark commented on kernel.go:
```diff
@@ -389,8 +398,8 @@ func handleExecuteRequest(ir *interp.Interp, receipt msgReceipt) error {
-		// if one or more value is image.Image, display it instead
-		vals = publishImages(vals, &receipt)
+		// if one or more value is image.Image or DisplayData, display it instead
+		vals = publishImagesAndDisplayData(vals, &receipt)
```
See the [Execution Result message format](http://jupyter-client.readthedocs.io/en/stable/messaging.html#id6) which is display data plus a bit extra. This is a minor detail but if you notice the `Out[_]` is missing when a value is returned because here it is being displayed rather than put in an execution result.

SpencerPark commented:
@cosmos72 that image is beautiful! I have the one minor comment above but that is looking really great.

SpencerPark commented:
Yes this was another thing Dan and I talked a lot about back when trying to design an API for this. 

The IPython semantics are such that when a single result is returned by the eval, that result is rendered nicely. If multiple then a string representation is returned. We opted for `fmt.Sprint`. Now to better suit the Go environment a `nil` result is considered non-existent and therefor situations that, for example, return an image and an error don't break the nice rendering by including `error` in their signature because when the error is `nil` we just render the image. 

Thanks for putting up with all these requests :)

cosmos72 commented:
You mean that you propose the following algorithm for values returned by the REPL ?

"if a single non-nil value is returned by REPL (other values, if, present, are nil), and such value is a DisplayData or an image.Image, it is displayed graphically. If multiple non-nil values are returned, they are displayed as string"

If that's what you propose, I politely disagree, because I find it confusing for users and unnecessarily limiting. My proposal, which is implemented in the current pull request, is:

"if one or more values returned by REPL is a DisplayData or an image.Image, it is displayed graphically and replaced with a placeholder (currently 'nil') in the returned values. Then the returned values are also passed as string to ExecutionResult, unless they are all nil or placeholders"

This has the advantage of not special-casing the single-value results, and it can display multiple images / html / ... returned by a single expression. If a user wants to see the string representation of a DisplayData or image.Image, IMHO it's quite straightforward for him/her to come up with the idea to call fmt.Sprint(val)

SpencerPark commented:
> "if a single non-nil value is returned by REPL (other values, if present, are nil), and such value is a DisplayData or an image.Image, it is displayed graphically. If multiple non-nil values are returned, they are displayed as string"

Correct, that is what I propose.

> Python does not have multiple return values, so any "multiple values" returned by an expression are actually wrapped in a tuple, list, dictionary or object - IMHO in such case it makes much more sense not to extract them for graphical displaying. Go multiple return values are thus a language feature with no Python counterpart, and should be considered when designing an API.

Sure the implementation is with tuples, but with syntax for building that tuple without parentheses it's really not fair to say Python doesn't have multiple return values. You can write code that assumes multiple return values without ever needing to know there was a tuple conversion in the middle. The fact that it is a tuple just makes it play nicer with Python.

The part that makes this approach a bit more Go friendly is considering `nil` as a non-returned value because return values are given a default value and with the Go pattern of returning an error in addition to a return value, you would never be able to display anything. The intent when returning an image and a `nil` error is that you are returning an image. This is captured in the semantics of the single non-nil return value giving the richer representation.

> "if one or more values returned by REPL is a DisplayData or an image.Image, it is displayed graphically and replaced with a placeholder (currently 'nil') in the returned values. Then the returned values are also passed as string to ExecutionResult, unless they are all nil or placeholders"

I disagree with this approach. There is a message in the protocol dedicated for cell results and this is the place it needs to be used. If multiple images (for example) need to be displayed then those need to be captured in variables and displayed separately. Since images can't be displayed side-by-side anyways the implicit transposition is also (in my opinion) not very natural.

The default behavior should always be a string representation. In some cases where we are sure the intent is something nicer then we can display that. If a string is a return value there is no doubt that it would be returned (in an `execution_result`) and not _displayed_ on stdout for example. The display _stream_ should be treated the same way. It needs to be used explicitly.

cosmos72 commented:
I am not convinced, but for the sake of finding an agreement I modified the pull request as you asked:
automatically display image.Image or display.Data returned by REPL only if it's the only non-nil result

SpencerPark commented:
I don't want a disagreement to hold up this feature. You know Go very well and so if you are strong on this then I think we should bring it into gophernotes.

> automatically display image.Image or display.Data returned by REPL only if it's the only non-nil result

One last note on this last commit: if we are going with the single output approach it should be put into an `execute_result` as otherwise the restriction for 1 output is unjustified. Or we can keep sending `display_data` messages and then might as well go back to your approach of supporting multiple returns.

On an unrelated note, thanks for all the work lately. You are bringing life back into this project and it's really great to see.

cosmos72 commented:
Ah, ok. I read your comment previously about using `execute_result` instead of `display_data` but I did not understand what you meant. Now it's clear, thanks.

Well, I did not want to get stuck in a discussion "A: let's do my way. B: No, let's do my way", and I also don't want to get stuck in the opposite "A: let's do your way. B: No, let's do your way"

The single-result functions are almost ready, let's use them. In case, we can extend to multiple results later.

cosmos72 commented:
modified to use `execute_result` instead of `display_data` if there's a single image.Image or Data returned by REPL

cosmos72 commented:
merged
